bigbang/ingress/__init__.py

```diff
@@ -0,0 +1,15 @@
+from bigbang.ingress.abstract import (
+    AbstractArchive,
+    AbstractList,
```

Comment:
I wonder why there are both general Message/List/Archive classes in bigbang.bigbang_io and also Abstract versions of the same in bigbang.ingress.abstract

Would it be possible for the ListServe and W3C classes to be subclasses of the bigbang.bigbang_io classes? Or is the Abstract class architecture really necessary for some reason?

Reply:
I wrote bigbang.bigbang_io to avoid duplicating methods that are concerned with reading/writing a message, list, or archive from/to various file formats. These function are generally needed in all parts (ingress, analysis and visualising) of BigBang.

At the moment I see the following complications with making, e.g., `bigbang.ingress.W3CList` a subclass of `ListIO`:
1) we can't led `AbstractList` inheret from `ListIO` because `@classmethod`'s from which the classes is initialised need to be part of the class itself (and not the one from which they inherit).
2) `W3CList.from_mbox()` returns something different than ListIO.from_mbox(). This is because `W3CList.from_mbox()` initialises a class that has further attributes (such as `filepath` and `name`) while `ListIO.from_mbox()` simply reads the content of the file at `filepath` and returns it as a list.

Reply:
Ok. If I understand correctly, then it sounds like ListIO is just a bundle of email related IO methods, not something that's meant to be used in an object oriented way.

Ok, so I would maintain that the current version of this code is confusing for a couple reasons:
 - there are many other classes in this proposal named "List" or some variation, which function very differently from ListIO. There's no parallelism here.
 - `ListIO.from_mbox()` returns a `list` which is different from an instantiation of a subclass of `AbstractList`...

I think some of the following changes could remedy this confusion.

 - Changing ListIO.from_mbox() into a standalone function, `list_from_mbox()`. (I.e. abandon OOP for a 'utility function' model), etc.
 - If maintaining the object-oriented design is necessary here, make these methods `@classmethods`
 - When a generic Python data type is being systematically used to represent mailing list data, use either [New Types](https://docs.python.org/3/library/typing.html#newtype) or [Type Aliases](https://docs.python.org/3/library/typing.html#type-aliases) to wrap those data structures in a named way.

I haven't used the Python typing library very much myself, but I like the way you've introduced it and I think it would improve the library to make better and further use of it.

Reply:
I have changed all methods in `bigbang/bigbang_io.py` into standalone functions and introduced costume data types which are defined in `bigbang/data_types.py`